iceoryx_utils/include/iceoryx_utils/internal/units/duration.inl

```diff
@@ -20,262 +20,511 @@ namespace iox
+    // one bit of the result represents 2^32 nanoseconds;
+    // just shifting left by 32 bits would result in an overflow, therefore blocks of full seconds must be extracted of
+    // the result;
+    // this cannot be done by dividing through NANOSECS_PER_SEC, since that one is base 1_000_000_000 and the result is
+    // base 2^32, therefore the least common multiple must be used to get blocks of full seconds represented as 2^32
+    // nanoseconds per bit;
+    // this can then safely be converted to seconds as well as nanoseconds without loosing precision
+
+    // least common multiple of 2^32 and NANOSECONDS_PER_SECOND
+    constexpr uint64_t LEAST_COMMON_MULTIPLE{8388608000000000};
+    static_assert(LEAST_COMMON_MULTIPLE % (1ULL << 32) == 0, "invalid multiple");
+    static_assert(LEAST_COMMON_MULTIPLE % NANOSECS_PER_SEC == 0, "invalid multiple");
+
+    constexpr uint64_t ONE_FULL_BLOCK_OF_SECONDS_ONLY{LEAST_COMMON_MULTIPLE >> 32};
+    constexpr uint64_t SECONDS_PER_FULL_BLOCK{LEAST_COMMON_MULTIPLE / NANOSECS_PER_SEC};
+
+    uint64_t multiplicatorHigh = static_cast<uint32_t>(multiplicator >> 32U);
+    auto nanosecondsFromHigh = m_nanoseconds * multiplicatorHigh;
+    auto fullBlocksOfSecondsOnly = nanosecondsFromHigh / ONE_FULL_BLOCK_OF_SECONDS_ONLY;
+    auto remainingBlockWithFullAndFractionalSeconds = nanosecondsFromHigh % ONE_FULL_BLOCK_OF_SECONDS_ONLY;
+
+    auto durationFromNanosecondsHigh = Duration{fullBlocksOfSecondsOnly * SECONDS_PER_FULL_BLOCK, 0U}
+                                       + Duration::nanoseconds(remainingBlockWithFullAndFractionalSeconds << 32);
+
+    return durationFromSeconds + durationFromNanosecondsLow + durationFromNanosecondsHigh;
 }
 
-inline constexpr Duration operator"" _us(unsigned long long int value) // PRQA S 48
+
+template <typename From, typename To>
+inline constexpr bool Duration::wouldCastFromFloatingPointProbablyOverflow(const From floatingPoint) const noexcept
 {
-    return Duration{static_cast<long double>(value) / 1000000.0};
+    static_assert(std::is_floating_point<From>::value, "only floating point is allowed");
+
+    // depending on the internal representation this could be either the last value to not cause an overflow
+    // or the first one which causes an overflow;
+    // to be safe, this is handled like causing an overflow which would result in undefined behavior when casting to
+    // Seconds_t
+    constexpr To SECONDS_BEFORE_LIKELY_OVERFLOW = static_cast<To>(std::numeric_limits<Seconds_t>::max());
+    return floatingPoint >= SECONDS_BEFORE_LIKELY_OVERFLOW;
 }
 
-inline constexpr Duration operator"" _ms(long double value)
+template <typename T>
+inline constexpr Duration Duration::fromFloatingPointSeconds(const T floatingPointSeconds) const noexcept
 {
-    return Duration{value / 1000.0};
+    static_assert(std::is_floating_point<T>::value, "only floating point is allowed");
+
+    if (std::isinf(floatingPointSeconds))
```

Comment:
Mmhhh. What if Duration == 0? Then you have 0 * infinity? This could become really really weird.

But we get in either case a warning!

Reply:
Well, I decided by a coin flip. Will change it to 0 if Duration is 0